channeldb/graph.go

```diff
@@ -236,6 +236,93 @@ func (c *ChannelGraph) ForEachChannel(cb func(*ChannelEdgeInfo, *ChannelEdgePoli
 	})
 }
 
+// ForEachNodeChannel iterates through all channels of a given node, executing the
+// passed callback with an edge info structure and the policies of each end
+// of the channel. The first edge policy is the outgoing edge *to* the
+// the connecting node, while the second is the incoming edge *from* the
+// connecting node. If the callback returns an error, then the iteration is
+// halted with the error propagated back up to the caller.
+//
+// Unknown policies are passed into the callback as nil values.
+//
+// The caller must provide an existing boltdb transaction.
+func (c *ChannelGraph) ForEachNodeChannel(node []byte, tx *bbolt.Tx,
+	cb func(*ChannelEdgeInfo, *ChannelEdgePolicy, *ChannelEdgePolicy) error) error {
+
+	traversal := func(tx *bbolt.Tx) error {
+		nodes := tx.Bucket(nodeBucket)
+		if nodes == nil {
+			return ErrGraphNotFound
+		}
+		edges := tx.Bucket(edgeBucket)
+		if edges == nil {
+			return ErrGraphNotFound
+		}
+		edgeIndex := edges.Bucket(edgeIndexBucket)
+		if edgeIndex == nil {
+			return ErrGraphNoEdgesFound
+		}
+
+		// In order to reach all the edges for this node, we take
+		// advantage of the construction of the key-space within the
+		// edge bucket. The keys are stored in the form: pubKey ||
+		// chanID. Therefore, starting from a chanID of zero, we can
+		// scan forward in the bucket, grabbing all the edges for the
+		// node. Once the prefix no longer matches, then we know we're
+		// done.
+		var nodeStart [33 + 8]byte
+		copy(nodeStart[:], node)
+		copy(nodeStart[33:], chanStart[:])
+
+		// Starting from the key pubKey || 0, we seek forward in the
+		// bucket until the retrieved key no longer has the public key
+		// as its prefix. This indicates that we've stepped over into
+		// another node's edges, so we can terminate our scan.
+		edgeCursor := edges.Cursor()
+		for nodeEdge, _ := edgeCursor.Seek(nodeStart[:]); bytes.HasPrefix(nodeEdge, node); nodeEdge, _ = edgeCursor.Next() {
+			// If the prefix still matches, the channel id is
+			// returned in nodeEdge. Channel id is used to lookup
+			// the node at the other end of the channel and both
+			// edge policies.
+			chanID := nodeEdge[33:]
+			edgeInfo, err := fetchChanEdgeInfo(edgeIndex, chanID)
+			if err != nil {
+				return err
+			}
+			edgeInfo.db = c.db
+
+			outgoingPolicy, err := fetchChanEdgePolicy(
+				edges, chanID, node, nodes,
+			)
+			if err != nil {
+				return err
+			}
+
+			otherNode, err := edgeInfo.OtherNodeKeyBytes(node)
+			if err != nil {
+				return err
+			}
+
+			incomingPolicy, err := fetchChanEdgePolicy(
+				edges, chanID, otherNode, nodes,
+			)
+			if err != nil {
+				return err
+			}
+
+			err = cb(&edgeInfo, outgoingPolicy, incomingPolicy)
+			if err != nil {
+				return err
+			}
```

Comment:
Can this code be reused from `LightningNode.ForEachChannel`? Possibly that function calling into here.

Reply:
I think `ForEachNodeChannel` doesn't even need to be used.  If we were to create a `LightningNode` struct containing only the public key and then call `ForEachChannel`, I think it would still have the same effect? We don't need to add the `db` field either, since the `ChannelEdgeInfo` never uses it in this context.

Reply:
It is a bit a matter of personal preference, but I normally try not to use structs in a way they were not intended to be used. If a function only accepts the parameters (including the struct fields it is a method of) it really requires, it is easier for others to understand its behavior. 

But in any case, copy/paste doesn't seem to be necessary.

Reply:
Gotcha, I will move the `traversal` outside of `ForEachChannel` and have it accept a boltdb transaction, public key bytes, and the db field.  That should dedupe the code well I think?

Reply:
I actually cannot do what I suggested above because of the traversal closure is used here:

https://github.com/lightningnetwork/lnd/blob/642fed53a3f0cefc750d79eaf0ae1d31d2217381/channeldb/graph.go#L2349-L2351

`View` takes a `func(*Tx) error)`, so the closure is actually necessary and can't really be moved out of the function.

Reply:
Done